### PR TITLE
Feature/ghc gmp

### PR DESCRIPTION
**Pull request by angerman**
###### Motivation for this change

Broken out of #43559

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/compilers/ghc/8.4.3.nix [hunk 1]
```diff
@@ -149,9 +150,9 @@ stdenv.mkDerivation (rec {
   configureFlags = [
     "--datadir=$doc/share/doc/ghc"
     "--with-curses-includes=${ncurses.dev}/include" "--with-curses-libraries=${ncurses.out}/lib"
-  ] ++ stdenv.lib.optional (targetPlatform == hostPlatform && ! enableIntegerSimple) [
-    "--with-gmp-includes=${gmp.dev}/include" "--with-gmp-libraries=${gmp.out}/lib"
-  ] ++ stdenv.lib.optional (targetPlatform == hostPlatform && hostPlatform.libc != "glibc" && !targetPlatform.isWindows) [
+  ] ++ stdenv.lib.optional (targetPlatform == hostPlatform && !enableIntegerSimple) [
+    "--with-gmp-includes=${targetPackages.gmp.dev}/include" "--with-gmp-libraries=${targetPackages.gmp.out}/lib"
+  ]) ++ stdenv.lib.optional (targetPlatform == hostPlatform && hostPlatform.libc != "glibc" && !targetPlatform.isWindows) [
```
Comment (matthewbauer) [inline pkgs/development/compilers/ghc/8.4.3.nix]: The extra ```)``` appears to be a typo

Comment (angerman) [reply pkgs/development/compilers/ghc/8.4.3.nix]: Sorry, fat-fingered that :(

**Comment by peti** [comment]
How did you test these changes? 

Do `ghcWithPackages` and `ghcWithHoogle` still work properly?

**Comment by angerman** [comment]
I've only tested this with a mingw cross compiler. That is, I built the full ghc on the host:macOS, and a cross compiler:macOS -> mingw.

GHC needs one integer library be it integer-simple or integer-gmp.
When providing the GMP library, it needs to be the target one (as @Ericson2314 pointed out, `targetPackages` is essentially `id` when build == host == target).

GHC also has an`--in-tree-gmp` option, which is usually the fallback if no GMP library is found.

**Review by matthewbauer — APPROVED** [review]

**Review by Ericson2314 — APPROVED** [review]

**Comment by Ericson2314** [comment]
@peti Err the point of this is the GMP ones, not the integer simple ones. GMP was just being inferred by default before, and that inferred (on the GHC side) happened to be unreliable (in the cross case). This forces the build system to use GMP when we want it to. By contrast, I'm not sure why anything has changed for the integer-simple builds, unless those happen to depend on the GMP ones for some reason and thus are downstream rebuilds.

**Comment by FRidh** [comment]
@matthewbauer Please use staging in the future.

**Comment by matthewbauer** [comment]
Ok - will do that next time. Is that recommended for GHC though? Basically any change will be 500+ (no stdenv changes).

**Comment by domenkozar** [comment]
I think it's best to use https://discourse.nixos.org/t/the-haskell-infrastructure-needs-new-maintainers/545/11?u=domenkozar

**Comment by matthewbauer** [comment]
Ok sounds good. Sorry for any inconvenience.

**Pull request merged**